server/snapshot.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2017 Apcera Inc. All rights reserved.
+
+package server
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/hashicorp/raft"
+	"github.com/nats-io/go-nats-streaming/pb"
+)
+
+// channelSnapshot implements the raft.FSMSnapshot interface.
+type channelSnapshot struct {
+	channel *channel
+}
+
+func newChannelSnapshot(c *channel) raft.FSMSnapshot {
+	return &channelSnapshot{channel: c}
+}
+
+// Persist should dump all necessary state to the WriteCloser 'sink',
+// and call sink.Close() when finished or call sink.Cancel() on error.
+func (c *channelSnapshot) Persist(sink raft.SnapshotSink) (err error) {
+	// TODO: this is very expensive and might repeatedly fail if messages are
+	// constantly being truncated. Is there a way we can optimize this, e.g.
+	// handling Restore() out-of-band from Raft?
+
+	defer func() {
+		if err != nil {
+			sink.Cancel()
+		}
+	}()
+
+	var (
+		first uint64
+		last  uint64
+		msg   *pb.MsgProto
+		data  []byte
+	)
+	first, last, err = c.channel.store.Msgs.FirstAndLastSequence()
+	if err != nil {
+		return err
+	}
+
+	buf := make([]byte, 4)
+	for seq := first; seq <= last; seq++ {
+		msg, err = c.channel.store.Msgs.Lookup(seq)
+		if err != nil {
+			return
+		}
+		if msg == nil {
+			// Channel truncation has occurred while snapshotting.
+			err = errors.New("channel was truncated while snapshotting")
+			return
+		}
```

Comment:
Improve tracing and code coverage with this code instead:
```
// If msg is nil, channel truncation has occurred while snapshotting.
if msg == nil {
    return errors.New("channel %q was truncated while snapshotting", c.channel.name)
}
```
Since you have named the returned error, any error that you return can be checked in the defer function.
In other words, say that you had named it `.. (retError error)`, then in your function you don't have to use `retErr =`, you can use local `err` if you want. If the function does return an error, you can check `if retError != nil` in the exit defer function.
That is, you could have:
```
defer func() {
   if retErr != nil {
      sink.Cancel()
   }
}()
...
if err := something(); err != nil {
   return err
}
```
This would cause the `if` statement in the `defer` function to evaluate to true.